### PR TITLE
PSP-8316 | Lease consultation updates

### PR DESCRIPTION
**Pull request by FuriousLlama**
Added view and containers on the front end for consultations. Updated backend for new specifications

image

### CODE REVIEW

**Comment by devinleighsmith** [comment]
The story, mockup, and confluence seem to have "There are no approvals or consultations." not "There are no consultations."

**Comment by devinleighsmith** [comment]
According to the mockup/confluence the heading should be Approvals/Consultation (although I wonder if it should be Approvals/Consultations)

**Comment by devinleighsmith** [comment]
Oh, and the button should be Add Approval/Consultation apparently.

I'm getting the feeling that maybe the mockups/confluence changed during development?

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/EditConsultationYupSchema.ts
Comment (devinleighsmith) [inline source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/EditConsultationYupSchema.ts:13]: I think this text is probably not correct for this field.

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/EditConsultationYupSchema.ts
Comment (devinleighsmith) [inline source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/EditConsultationYupSchema.ts:9]: Perhaps: Other description?

**Comment by devinleighsmith** [comment]
Oddly, I don't see any documentation stating that Response received on is conditionally rendered, but that is the current logic. Is there a requirement I'm not seeing.

One issue with the current implementation is that if I set Response received Yes, and set a date, and then set response received No, the date is still set, and is displayed on the view only version of the screen, even when set to No.

**Comment by devinleighsmith** [comment]
@FuriousLlama what is the plan for the DB issue related to this story? Does Doug have a defect logged? my concern is that the next sprint is a release sprint, and I'm not sure Doug has made any corrections for this issue in 88.

**Comment by devinleighsmith** [comment]
Consultation type on the add/edit/view form should be Consultation/Approval type: or Approval/Consultation type depending on confluence or the mockup

**Comment by devinleighsmith** [comment]
There are a number of tooltips that were requested on the create/edit screen that appear to be missing (mentioned on the mockup and confluence).

**Comment by devinleighsmith** [comment]
Hmm, the confluence page specifies that the only allowed values for the Response received or Yes/No but the implementation has Unknown as well.

**Comment by devinleighsmith** [comment]
The text on the modal is Approval/Consultation as well.

**Comment by devinleighsmith** [comment]
Hmm, also I think the Consultation top-level header is being styled the same as the other headers. In the mockup I think it is using different styling (Hopefully consistent with existing PSP headers).

File: source/backend/api/Areas/Leases/Controllers/ConsultationController.cs
Comment (devinleighsmith) [inline source/backend/api/Areas/Leases/Controllers/ConsultationController.cs:82]: LeaseEdit?

File: source/backend/api/Areas/Leases/Controllers/ConsultationController.cs
Comment (devinleighsmith) [inline source/backend/api/Areas/Leases/Controllers/ConsultationController.cs:176]: not sure if it matters, but:
1) the lease id is unused.
2) the other endpoints validate that the lease id passed is valid for the given consultationId.

File: source/backend/api/Services/LeaseService.cs
Comment (devinleighsmith) [inline source/backend/api/Services/LeaseService.cs:360]: unused param.

File: source/backend/apimodels/Models/Concepts/Lease/ConsultationLeaseMap.cs
Comment (devinleighsmith) [inline source/backend/apimodels/Models/Concepts/Lease/ConsultationLeaseMap.cs:42]: is this a DateOnly or a DateTime?

File: source/backend/dal/Repositories/ConsultationRepository.cs
Comment (devinleighsmith) [inline source/backend/dal/Repositories/ConsultationRepository.cs:40]: Do we need the ConsultationStatusTypeCodeNavigation as well? or neither?

File: source/frontend/src/components/common/Section/Section.tsx
Comment (devinleighsmith) [inline source/frontend/src/components/common/Section/Section.tsx:33]: this was a strange one.

File: source/frontend/src/components/common/buttons/Button.tsx
Comment (devinleighsmith) [inline source/frontend/src/components/common/buttons/Button.tsx:69]: ooph, this seems scary, are you sure that this doesn't have any unintended consequences?

File: source/frontend/src/components/common/buttons/Button.tsx
Comment (devinleighsmith) [inline source/frontend/src/components/common/buttons/Button.tsx:79]: same here.

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/detail/ConsultationListView.tsx
Comment (devinleighsmith) [inline source/frontend/src/features/mapSideBar/lease/tabs/consultations/detail/ConsultationListView.tsx]: See github actions warnings about unused variables.

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/detail/ConsultationListView.tsx
Comment (devinleighsmith) [inline source/frontend/src/features/mapSideBar/lease/tabs/consultations/detail/ConsultationListView.tsx:65]: You probably couldn't see this, because load times are so fast on our locals, but this is what the loading screen looks like:

image

My preference for these is to render an empty version of the screen with the loading backdrop overtop, but some kind of correction required here.

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/detail/ConsultationListView.tsx
Comment (devinleighsmith) [inline source/frontend/src/features/mapSideBar/lease/tabs/consultations/detail/ConsultationListView.tsx:108]: is null/undefined checking required here? aka ?? ''

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationAddContainer.tsx
Comment (devinleighsmith) [inline source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationAddContainer.tsx:45]: hmm, the only issue I see is that if somehow this is false, no error is presented but also the onSuccess and navigation isn't fired.

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationAddContainer.tsx
Comment (devinleighsmith) [inline source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationAddContainer.tsx:47]: Should this component be handling navigation or should it just be firing callbacks?

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationEditForm.tsx
Comment (devinleighsmith) [inline source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationEditForm.tsx:67]: I think these forms could use the <UnsavedChangesPrompt /> component as well.

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationUpdateContainer.tsx
Comment (devinleighsmith) [inline source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationUpdateContainer.tsx]: some more unused variables detected by GH.

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationUpdateContainer.tsx
Comment (devinleighsmith) [inline source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationUpdateContainer.tsx:60]: typo, should be organization not person.

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/models.ts
Comment (devinleighsmith) [inline source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/models.ts:62]: hmm, I think this is ok - just want to make sure we shouldn't be conditionally setting this based on whether the id is an org?

File: source/frontend/src/hooks/pims-api/useApiConsultations.ts
Comment (devinleighsmith) [inline source/frontend/src/hooks/pims-api/useApiConsultations.ts:18]: is agreementId correct?

File: source/frontend/src/hooks/repositories/useConsultationProvider.ts
Comment (devinleighsmith) [inline source/frontend/src/hooks/repositories/useConsultationProvider.ts:63]: Hmm, I think you have try/catches down the line from these calls, but I think the default behaviour is to catch errors from these and use the onError provided below. If you want to throw an exception I think you need throwError: true.

File: source/frontend/src/interfaces/IContactSearchResult.ts
Comment (devinleighsmith) [inline source/frontend/src/interfaces/IContactSearchResult.ts:57]: thanks for this.

**Review by devinleighsmith — CHANGES_REQUESTED** [review]
Only blocking this because I'm concerned about the changes to the base button styling. If you are confident that isn't an issue, let me know.

**Comment by FuriousLlama** [comment]
> Only blocking this because I'm concerned about the changes to the base button styling. If you are confident that isn't an issue, let me know.

Pretty confident is not an issue. That extra padding was causing alignment issues all over the app. Checked when it was added and its very early on so I don't think it was part of any button styling enhancement.

**Comment by FuriousLlama** [comment]
> Oddly, I don't see any documentation stating that Response received on is conditionally rendered, but that is the current logic. Is there a requirement I'm not seeing.
> 
> One issue with the current implementation is that if I set Response received Yes, and set a date, and then set response received No, the date is still set, and is displayed on the view only version of the screen, even when set to No.

Yes, I infered that. I will check with dev once this hits DEV and adjust it as necessary

File: source/backend/apimodels/Models/Concepts/Lease/ConsultationLeaseMap.cs
Comment (FuriousLlama) [reply source/backend/apimodels/Models/Concepts/Lease/ConsultationLeaseMap.cs:42]: Its a datetime, however we only need dates, that's wh y the model changes it

File: source/backend/dal/Repositories/ConsultationRepository.cs
Comment (FuriousLlama) [reply source/backend/dal/Repositories/ConsultationRepository.cs:40]: Consultation status is no longer used.

File: source/frontend/src/components/common/Section/Section.tsx
Comment (FuriousLlama) [reply source/frontend/src/components/common/Section/Section.tsx:33]: yeah, changed it after some tests to make sure it does what we want

File: source/frontend/src/components/common/buttons/Button.tsx
Comment (FuriousLlama) [reply source/frontend/src/components/common/buttons/Button.tsx:69]: Pretty confident, was causing misalignment on a  lot of pages

File: source/frontend/src/components/common/buttons/Button.tsx
Comment (FuriousLlama) [reply source/frontend/src/components/common/buttons/Button.tsx:79]: same as above

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/detail/ConsultationListView.tsx
Comment (FuriousLlama) [reply source/frontend/src/features/mapSideBar/lease/tabs/consultations/detail/ConsultationListView.tsx]: I think you have seen an old commit?

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/detail/ConsultationListView.tsx
Comment (FuriousLlama) [reply source/frontend/src/features/mapSideBar/lease/tabs/consultations/detail/ConsultationListView.tsx:65]: I think there is some room for improvement here. Will chat win ana to check what her thoughts are on loading.

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/detail/ConsultationListView.tsx
Comment (FuriousLlama) [reply source/frontend/src/features/mapSideBar/lease/tabs/consultations/detail/ConsultationListView.tsx:108]: Not needed, the view model uses empty strings if the value coming from the wire is empty

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationAddContainer.tsx
Comment (FuriousLlama) [reply source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationAddContainer.tsx:45]: That is true, however, that's our current pattern in other screens.

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationAddContainer.tsx
Comment (FuriousLlama) [reply source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationAddContainer.tsx:47]: Navigation at the container makes sense to me.

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationEditForm.tsx
Comment (FuriousLlama) [reply source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationEditForm.tsx:67]: Not sure what you mean here.

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationUpdateContainer.tsx
Comment (FuriousLlama) [reply source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationUpdateContainer.tsx]: I think this is an earlier commit

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationUpdateContainer.tsx
Comment (FuriousLlama) [reply source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationUpdateContainer.tsx:90]: Same as above

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/models.ts
Comment (FuriousLlama) [reply source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/models.ts:62]: AFAIK only persons can be primary contacts

File: source/frontend/src/hooks/repositories/useConsultationProvider.ts
Comment (FuriousLlama) [reply source/frontend/src/hooks/repositories/useConsultationProvider.ts:63]: thats right. I've updated it now

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/detail/ConsultationListView.tsx
Comment (devinleighsmith) [reply source/frontend/src/features/mapSideBar/lease/tabs/consultations/detail/ConsultationListView.tsx:108]: This is what I'm seeing locally:

image

File: source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationEditForm.tsx
Comment (devinleighsmith) [reply source/frontend/src/features/mapSideBar/lease/tabs/consultations/edit/ConsultationEditForm.tsx:67]: Oh, one of my comments must not have been saved, referring to the ConfirmNavigation component.

File: source/frontend/src/components/common/buttons/Button.tsx
Comment (devinleighsmith) [reply source/frontend/src/components/common/buttons/Button.tsx:69]: Its subtle, but this has definitely re-introduced the issue where the save and cancel buttons aren't the same size:

image

Could you either revert this, or find a way of sizing the cancel button (with its additional border) the same as the save button?

**Comment by devinleighsmith** [comment]
I had a look at the button styling, and this did introduce an issue, but its minor enough that I'll lift the change request. Would still prefer a solution before this is merged.

**Review by devinleighsmith — APPROVED** [review]
Approved, but would like some more changes before merge. Also assuming test coverage is coming in future?

**Pull request merged**
